src/ring_theory/derivations.lean

```diff
@@ -0,0 +1,200 @@
+  f (r • m) = r • f m :=
+by rw [compatible_smul A r m, linear_map.map_smul, ←compatible_smul A r (f m)]
+
+instance : has_coe (M →ₗ[A] N) (M →ₗ[R] N) :=
+⟨λ f, ⟨f.to_fun, λ x y, f.map_add' x y, λ r n, compatible_map_smul _ _ _⟩⟩
+
+end compatible_semimodule
+
+/-- Derivations in commutative algebra. To be changed.-/
+@[protect_proj]
+structure derivation (R : Type*) (A : Type*) [comm_semiring R] [comm_semiring A]
+  [algebra R A] (M : Type*) [add_cancel_comm_monoid M] [semimodule A M] [semimodule R M]
+  [compatible_semimodule R A M]
+  extends A →ₗ[R] M :=
+(leibniz' (a b : A) : to_fun (a * b) = a • to_fun b + b • to_fun a)
+
+namespace derivation
+
+section
+
+variables {R : Type*} [comm_semiring R]
+{A : Type*} [comm_semiring A] [algebra R A]
+{M : Type*} [add_cancel_comm_monoid M] [semimodule A M] [semimodule R M]
+[compatible_semimodule R A M]
+(D : derivation R A M) {D1 D2 : derivation R A M} (r : R) (a b : A)
+
+instance : has_coe_to_fun (derivation R A M) := ⟨_, λ D, D.to_linear_map.to_fun⟩
+
+instance has_coe_to_linear_map : has_coe (derivation R A M) (A →ₗ[R] M) :=
+  ⟨λ D, D.to_linear_map⟩
+
+@[simp, norm_cast]
+lemma coe_linear_map (f : derivation R A M) :
+  ⇑(f : A →ₗ[R] M) = f := rfl
+
+lemma coe_injective (H : ⇑D1 = D2) : D1 = D2 :=
+by cases D1; cases D2; congr'; exact linear_map.coe_inj H
+
+@[ext] theorem ext (H : ∀ a, D1 a = D2 a) : D1 = D2 :=
+coe_injective $ funext H
+
+@[simp] lemma map_add : D (a + b) = D a + D b := is_add_hom.map_add D a b
+@[simp] lemma map_zero : D 0 = 0 := is_add_monoid_hom.map_zero D
+@[simp] lemma leibniz : D (a * b) = a • D b + b • D a := D.leibniz' _ _
+@[simp] lemma map_smul : D (r • a) = r • D a := linear_map.map_smul D r a
+
+@[simp] lemma map_one_eq_zero : D 1 = 0 :=
+begin
+  have h : D 1 = D (1 * 1) := by rw mul_one,
+  rw [leibniz D 1 1, one_smul] at h, /- better way to do this? -/
+  exact zero_left_cancel h,
+end
+
+@[simp] lemma map_algebra_map : D (algebra_map R A r) = 0 :=
+by rw [←mul_one r, ring_hom.map_mul, map_one, ←smul_def, map_smul, map_one_eq_zero, smul_zero]
+
+instance : has_zero (derivation R A M) :=
+⟨⟨(0 : A →ₗ[R] M), λ a b, by simp only [add_zero, linear_map.zero_apply,
+                                      linear_map.to_fun_eq_coe, smul_zero]⟩⟩
+
+instance : inhabited (derivation R A M) := ⟨0⟩
+
+instance : add_comm_monoid (derivation R A M) :=
+{ add := λ D1 D2, ⟨D1 + D2, λ a b, begin
+  simp only [leibniz, linear_map.add_apply, linear_map.to_fun_eq_coe, coe_linear_map, smul_add],
+  cc, end⟩,
+  add_assoc := λ D E F, ext $ λ a, add_assoc _ _ _,
+  zero_add := λ D, ext $ λ a, zero_add _,
+  add_zero := λ D, ext $ λ a, add_zero _,
+  add_comm := λ D E, ext $ λ a, add_comm _ _,
+  ..derivation.has_zero }
+
+@[priority 100]
+instance derivation.Rsemimodule : semimodule R (derivation R A M) :=
+{ smul := λ r D, ⟨r • D, λ a b, by simp only [linear_map.smul_apply, leibniz,
+linear_map.to_fun_eq_coe, compatible_smul_comm, coe_linear_map, smul_add, add_comm],⟩,
```

Comment:
```suggestion
{ smul := λ r D, ⟨r • D, λ a b, by simp only [linear_map.smul_apply, leibniz,
    linear_map.to_fun_eq_coe, compatible_smul_comm, coe_linear_map, smul_add, add_comm]⟩,
```
Did you also consider only having `by simp [smul_add, add_comm]`?

Reply:
Doesn't it make things longer to compile?

Reply:
True, without the `only` it is noticeably slower here.